test/general/test_entrance_rando.py

```diff
@@ -0,0 +1,376 @@
+                                      region_type: Type[Region] = Region):
+    """
+    Generates a grid-like region structure for ER testing, where menu is connected to the top-left region, and each
+    region "in vanilla" has 2 2-way exits going either down or to the right, until reaching the goal region in the
+    bottom right
+    """
+    for row in range(grid_side_length):
+        for col in range(grid_side_length):
+            index = row * grid_side_length + col
+            name = f"region{index}"
+            region = region_type(name, 1, multiworld)
+            multiworld.regions.append(region)
+            generate_locations(region_size, 1, region=region, tag=f"_{name}")
+
+            if row == 0 and col == 0:
+                multiworld.get_region("Menu", 1).connect(region)
+            if col != 0:
+                generate_entrance_pair(region, "_left", ERTestGroups.LEFT)
+            if col != grid_side_length - 1:
+                generate_entrance_pair(region, "_right", ERTestGroups.RIGHT)
+            if row != 0:
+                generate_entrance_pair(region, "_top", ERTestGroups.TOP)
+            if row != grid_side_length - 1:
+                generate_entrance_pair(region, "_bottom", ERTestGroups.BOTTOM)
+
+
+class TestEntranceLookup(unittest.TestCase):
+    def test_shuffled_targets(self):
+        """tests that get_targets shuffles targets between groups when requested"""
+        multiworld = generate_test_multiworld()
+        generate_disconnected_region_grid(multiworld, 5)
+
+        lookup = EntranceLookup(multiworld.per_slot_randoms[1], coupled=True)
+        er_targets = [entrance for region in multiworld.get_regions(1)
+                      for entrance in region.entrances if not entrance.parent_region]
+        for entrance in er_targets:
+            lookup.add(entrance)
+
+        retrieved_targets = lookup.get_targets([ERTestGroups.TOP, ERTestGroups.BOTTOM],
+                                               False, False)
+        prev = None
+        group_order = [prev := group.randomization_group for group in retrieved_targets if prev != group.randomization_group]
+        # technically possible that group order may not be shuffled, by some small chance, on some seeds. but generally
+        # a shuffled list should alternate more frequently which is the desired behavior here
+        self.assertGreater(len(group_order), 2)
+
+
+    def test_ordered_targets(self):
+        """tests that get_targets does not shuffle targets between groups when requested"""
+        multiworld = generate_test_multiworld()
+        generate_disconnected_region_grid(multiworld, 5)
+
+        lookup = EntranceLookup(multiworld.per_slot_randoms[1], coupled=True)
+        er_targets = [entrance for region in multiworld.get_regions(1)
+                      for entrance in region.entrances if not entrance.parent_region]
+        for entrance in er_targets:
+            lookup.add(entrance)
```

Comment:
This duplicated code could be in `setUp`, no?

Reply:
My opinion/experience with moving this kind of setup to a before-each-test setup method is that
1. It makes it difficult to follow what state is being tested, as your "arrange" is nowhere near your "act" and "assert" in source
2. As soon as you try to generalize the setup you will find many cases where it doesn't work well and you have to replicate the whole setup anyway

Since the duplicated block is small I believe duplicating it is cleaner, especially in a test scenario (in production code it might be different)